File: Validation/archive/temp_lua/find_duplicate_quests.lua
#!/usr/bin/env lua

-- Script to find duplicate quests in Questie database
-- Identifies quests with identical names and compares their data

-- Load the quest database
local function loadQuestDatabase()
    -- Define epochQuestData table
    epochQuestData = {}
    
    -- Load the database file
    dofile("Database/Epoch/epochQuestDB.lua")
    
    return epochQuestData
end

-- Compare two quest entries for equality
local function compareQuests(quest1, quest2)
    -- Check if both are tables
    if type(quest1) ~= "table" or type(quest2) ~= "table" then
        return false
    end
    
    -- Compare all 30 fields
    for i = 1, 30 do
        local v1 = quest1[i]
        local v2 = quest2[i]
        
        -- Handle nil values
        if v1 == nil and v2 == nil then
            -- Both nil, continue
        elseif v1 == nil or v2 == nil then
            -- One is nil, not equal
            return false
        elseif type(v1) ~= type(v2) then
            -- Different types
            return false
        elseif type(v1) == "table" then
            -- For tables, convert to string for comparison
            local function tableToString(t)
                if not t then return "nil" end
                local str = "{"
                for k, v in pairs(t) do
                    if type(v) == "table" then
                        str = str .. tableToString(v) .. ","
                    else
                        str = str .. tostring(v) .. ","
                    end
                end
                return str .. "}"
            end
            if tableToString(v1) ~= tableToString(v2) then
                return false
            end
        elseif v1 ~= v2 then
            -- Direct comparison for non-tables
            return false
        end
    end
    
    return true
end

-- Extract key quest data for comparison
local function getQuestSummary(quest)
    if not quest then return "nil" end
    
    local summary = {}
    
    -- Name (field 1)
    summary.name = quest[1] or "Unknown"
    
    -- Quest giver (field 2)
    if quest[2] and quest[2][1] then
        summary.startNpc = quest[2][1][1] or "none"
    else
        summary.startNpc = "none"
    end
    
    -- Quest turn-in (field 3)  
    if quest[3] and quest[3][1] then
        summary.endNpc = quest[3][1][1] or "none"
    else
        summary.endNpc = "none"
    end
    
    -- Objectives (field 10)
    summary.objectives = ""
    if quest[10] then
        -- Creatures to kill
        if quest[10][1] then
            for _, mob in ipairs(quest[10][1] or {}) do
                summary.objectives = summary.objectives .. string.format("Kill %dx%s; ", mob[2] or 0, mob[1] or "?")
            end
        end
        -- Objects to interact with
        if quest[10][2] then
            for _, obj in ipairs(quest[10][2] or {}) do
                summary.objectives = summary.objectives .. string.format("Object %dx%s; ", obj[2] or 0, obj[1] or "?")
            end
        end
        -- Items to collect
        if quest[10][3] then
            for _, item in ipairs(quest[10][3] or {}) do
                summary.objectives = summary.objectives .. string.format("Item %dx%s; ", item[2] or 0, item[1] or "?")
            end
        end
    end
    
    -- Zone (field 17)
    summary.zone = quest[17] or "none"
    
    -- Quest flags (field 23)
    summary.questFlags = quest[23] or 0
    
    return summary
end

-- Main function
local function findDuplicates()
    print("Loading quest database...")
    local quests = loadQuestDatabase()
    
    -- Build a map of quest names to IDs
    local nameToIds = {}
    local totalQuests = 0
    
    for questId, questData in pairs(quests) do
        totalQuests = totalQuests + 1
        local name = questData[1] or "Unknown"
        
        if not nameToIds[name] then
            nameToIds[name] = {}
        end
        table.insert(nameToIds[name], questId)
    end
    
    print(string.format("Analyzed %d quests", totalQuests))
    print("=" .. string.rep("=", 78))
    
    -- Find and analyze duplicates
    local duplicateGroups = {}
    local exactDuplicates = {}
    local similarQuests = {}
    
    for name, ids in pairs(nameToIds) do
        if #ids > 1 then
            -- Skip placeholder entries
            if not string.find(name, "^%[Epoch%] Quest %d+$") then
                table.insert(duplicateGroups, {name = name, ids = ids})
            end
        end
    end
    
    -- Sort by name for consistent output
    table.sort(duplicateGroups, function(a, b) return a.name < b.name end)
    
    -- Analyze each duplicate group
    print("\nQUESTS WITH DUPLICATE NAMES:")
    print("=" .. string.rep("=", 78))
    
    for _, group in ipairs(duplicateGroups) do
        print(string.format("\n\"%s\" - %d instances:", group.name, #group.ids))
        print("-" .. string.rep("-", 78))
        
        local questDataList = {}
        for _, id in ipairs(group.ids) do
            local quest = quests[id]
            local summary = getQuestSummary(quest)
            table.insert(questDataList, {id = id, quest = quest, summary = summary})
        end
        
        -- Check for exact duplicates
        local foundExactDuplicate = false
        for i = 1, #questDataList do
            for j = i + 1, #questDataList do
                if compareQuests(questDataList[i].quest, questDataList[j].quest) then
                    if not foundExactDuplicate then
                        print("  ⚠️  EXACT DUPLICATES FOUND:")
                    end
                    foundExactDuplicate = true
                    print(string.format("      Quest %d and %d are IDENTICAL - one can be purged", 
                        questDataList[i].id, questDataList[j].id))
                    
                    table.insert(exactDuplicates, {
                        name = group.name,
                        id1 = questDataList[i].id,
                        id2 = questDataList[j].id
                    })
                end
            end
        end
        
        -- Display quest details
        for _, data in ipairs(questDataList) do
            local s = data.summary
            print(string.format("  Quest %d:", data.id))
            print(string.format("    Start NPC: %s | End NPC: %s | Zone: %s | Flags: %s", 
                tostring(s.startNpc), tostring(s.endNpc), tostring(s.zone), tostring(s.questFlags)))
            if s.objectives ~= "" then
                print(string.format("    Objectives: %s", s.objectives))
            else
                print("    Objectives: None")
            end
        end
        
        -- Check if they're sequential (likely different parts of a chain)
        local sequential = true
        for i = 2, #group.ids do
            if group.ids[i] ~= group.ids[i-1] + 1 then
                sequential = false
                break
            end
        end
        
        if sequential and not foundExactDuplicate then
            print("  ℹ️  Note: Sequential IDs - likely a quest chain, should be preserved")
            table.insert(similarQuests, {name = group.name, ids = group.ids})
        end
    end
    
    -- Summary report
    print("\n" .. string.rep("=", 80))
    print("SUMMARY REPORT")
    print(string.rep("=", 80))
    print(string.format("Total quests analyzed: %d", totalQuests))
    print(string.format("Quest names with duplicates: %d", #duplicateGroups))
    print(string.format("Exact duplicates to purge: %d", #exactDuplicates))
    print(string.format("Similar quests to preserve (quest chains): %d", #similarQuests))
    
    -- List quests to purge
    if #exactDuplicates > 0 then
        print("\n" .. string.rep("=", 80))
        print("RECOMMENDED PURGE LIST:")
        print(string.rep("=", 80))
        print("The following quest IDs can be safely removed (keeping the lower ID):")
        
        for _, dup in ipairs(exactDuplicates) do
            local keepId = math.min(dup.id1, dup.id2)
            local removeId = math.max(dup.id1, dup.id2)
            print(string.format("  REMOVE Quest %d (duplicate of %d) - \"%s\"", 
                removeId, keepId, dup.name))
        end
    end
    
    -- List quest chains to preserve
    if #similarQuests > 0 then
        print("\n" .. string.rep("=", 80))
        print("QUEST CHAINS TO PRESERVE:")
        print(string.rep("=", 80))
        print("The following appear to be quest chains and should be kept:")
        
        for _, chain in ipairs(similarQuests) do
            local idStr = table.concat(chain.ids, ", ")
            print(string.format("  \"%s\": Quests %s", chain.name, idStr))
        end
    end
    
    -- Special case: Check for our known duplicate
    print("\n" .. string.rep("=", 80))
    print("SPECIFIC CASE: Shift into G.E.A.R.")
    print(string.rep("=", 80))
    if quests[28725] and quests[28901] then
        print("Quest 28725 exists: " .. (quests[28725][1] or "Unknown"))
        print("Quest 28901 exists: " .. (quests[28901][1] or "Unknown"))
        if compareQuests(quests[28725], quests[28901]) then
            print("✅ Confirmed: 28725 and 28901 are EXACT duplicates")
            print("Recommendation: Remove 28901, keep 28725 (actual quest ID)")
        else
            print("⚠️  Warning: 28725 and 28901 have the same name but different data")
            print("Further investigation needed")
        end
    end
end

-- Run the script
findDuplicates()

File: Validation/find_duplicate_quests.py
#!/usr/bin/env python3

"""
Script to find duplicate quests in Questie database
Identifies quests with identical names and compares their data
"""

import re
import json
from collections import defaultdict

def parse_lua_table(text):
    """Parse a Lua table into a Python structure."""
    # This is a simplified parser for the specific format we have
    result = {}
    
    # Find all quest entries like [questId] = {data}
    pattern = r'\[(\d+)\]\s*=\s*\{([^}]+(?:\{[^}]*\}[^}]*)*)\}'
    
    for match in re.finditer(pattern, text):
        quest_id = int(match.group(1))
        quest_data_str = match.group(2)
        
        # Parse the quest data (simplified - just extract what we need)
        # Extract the quest name (first quoted string)
        name_match = re.search(r'"([^"]*)"', quest_data_str)
        quest_name = name_match.group(1) if name_match else f"Quest {quest_id}"
        
        # Store the raw data for comparison
        result[quest_id] = {
            'name': quest_name,
            'raw_data': quest_data_str
        }
    
    return result

def extract_quest_details(raw_data):
    """Extract key details from quest data string."""
    details = {}
    
    # Split by commas but be careful with nested structures
    parts = []
    current = ""
    depth = 0
    
    for char in raw_data:
        if char == '{':
            depth += 1
        elif char == '}':
            depth -= 1
        elif char == ',' and depth == 0:
            parts.append(current.strip())
            current = ""
            continue
        current += char
    if current:
        parts.append(current.strip())
    
    # Field 23 is questFlags (0-indexed would be 22)
    if len(parts) > 22:
        # Extract questFlags value
        flags_str = parts[22].strip()
        try:
            details['questFlags'] = int(flags_str) if flags_str and flags_str != 'nil' else None
        except:
            details['questFlags'] = flags_str
    
    # Field 17 is zone
    if len(parts) > 16:
        zone_str = parts[16].strip()
        try:
            details['zone'] = int(zone_str) if zone_str and zone_str != 'nil' else None
        except:
            details['zone'] = zone_str
            
    return details

def find_duplicates():
    """Main function to find duplicate quests."""
    
    print("Loading quest database...")
    
    # Read the quest database file
    with open("Database/Epoch/epochQuestDB.lua", 'r', encoding='utf-8') as f:
        content = f.read()
    
    # Parse quests
    quests = parse_lua_table(content)
    print(f"Analyzed {len(quests)} quests")
    
    # Build a map of quest names to IDs
    name_to_ids = defaultdict(list)
    
    for quest_id, quest_data in quests.items():
        name = quest_data['name']
        # Skip placeholder entries
        if not name.startswith("[Epoch] Quest "):
            name_to_ids[name].append(quest_id)
    
    # Find duplicates
    duplicate_groups = []
    exact_duplicates = []
    
    for name, ids in name_to_ids.items():
        if len(ids) > 1:
            duplicate_groups.append((name, ids))
    
    # Sort by name for consistent output
    duplicate_groups.sort(key=lambda x: x[0])
    
    print("\n" + "=" * 80)
    print("QUESTS WITH DUPLICATE NAMES:")
    print("=" * 80)
    
    for name, ids in duplicate_groups:
        print(f"\n\"{name}\" - {len(ids)} instances:")
        print("-" * 78)
        
        # Check if raw data is identical
        quest_data_list = []
        for quest_id in ids:
            raw_data = quests[quest_id]['raw_data']
            details = extract_quest_details(raw_data)
            quest_data_list.append({
                'id': quest_id,
                'raw': raw_data,
                'details': details
            })
        
        # Check for exact duplicates
        found_exact = False
        for i in range(len(quest_data_list)):
            for j in range(i + 1, len(quest_data_list)):
                # Compare raw data (ignoring whitespace differences)
                data1 = re.sub(r'\s+', ' ', quest_data_list[i]['raw']).strip()
                data2 = re.sub(r'\s+', ' ', quest_data_list[j]['raw']).strip()
                
                if data1 == data2:
                    if not found_exact:
                        print("  ⚠️  EXACT DUPLICATES FOUND:")
                    found_exact = True
                    id1 = quest_data_list[i]['id']
                    id2 = quest_data_list[j]['id']
                    print(f"      Quest {id1} and {id2} are IDENTICAL - one can be purged")
                    exact_duplicates.append((name, id1, id2))
        
        # Display quest IDs and details
        for data in quest_data_list:
            quest_id = data['id']
            details = data['details']
            print(f"  Quest {quest_id}:")
            if 'questFlags' in details:
                print(f"    questFlags: {details['questFlags']}")
            if 'zone' in details:
                print(f"    Zone: {details['zone']}")
        
        # Check if sequential (likely a quest chain)
        sorted_ids = sorted(ids)
        sequential = all(sorted_ids[i] == sorted_ids[i-1] + 1 for i in range(1, len(sorted_ids)))
        
        if sequential and not found_exact:
            print("  ℹ️  Note: Sequential IDs - likely a quest chain, should be preserved")
    
    # Summary report
    print("\n" + "=" * 80)
    print("SUMMARY REPORT")
    print("=" * 80)
    print(f"Total quests analyzed: {len(quests)}")
    print(f"Quest names with duplicates: {len(duplicate_groups)}")
    print(f"Exact duplicates to purge: {len(exact_duplicates)}")
    
    # List quests to purge
    if exact_duplicates:
        print("\n" + "=" * 80)
        print("RECOMMENDED PURGE LIST:")
        print("=" * 80)
        print("The following quest IDs can be safely removed (keeping the lower ID):")
        
        for name, id1, id2 in exact_duplicates:
            keep_id = min(id1, id2)
            remove_id = max(id1, id2)
            print(f"  REMOVE Quest {remove_id} (duplicate of {keep_id}) - \"{name}\"")
    
    # Special case: Check for our known duplicate
    print("\n" + "=" * 80)
    print("SPECIFIC CASE: Shift into G.E.A.R.")
    print("=" * 80)
    
    if 28725 in quests and 28901 in quests:
        print(f"Quest 28725 exists: {quests[28725]['name']}")
        print(f"Quest 28901 exists: {quests[28901]['name']}")
        
        # Compare raw data
        data1 = re.sub(r'\s+', ' ', quests[28725]['raw_data']).strip()
        data2 = re.sub(r'\s+', ' ', quests[28901]['raw_data']).strip()
        
        if data1 == data2:
            print("✅ Confirmed: 28725 and 28901 are EXACT duplicates")
            print("Recommendation: Remove 28901, keep 28725 (actual quest ID)")
        else:
            print("⚠️  Warning: 28725 and 28901 have the same name but different data")
            print("Further investigation needed")

if __name__ == "__main__":
    find_duplicates()